module/apmfasthttp/utils.go

```diff
@@ -0,0 +1,114 @@
+package apmfasthttp
+
+import (
+	"net/http"
+	"net/url"
+
+	"github.com/savsgio/gotils/strconv"
+	"github.com/valyala/bytebufferpool"
+	"github.com/valyala/fasthttp"
+	"go.elastic.co/apm"
+	"go.elastic.co/apm/module/apmhttp"
+)
+
+func resetHTTPRequest(req *http.Request) {
+	req.Method = ""
+	req.URL = nil
+	req.Proto = ""
+	req.ProtoMajor = 0
+	req.ProtoMinor = 0
+	resetHTTPMap(req.Header)
+	req.Body = nil
+	req.GetBody = nil
+	req.ContentLength = 0
+	req.TransferEncoding = req.TransferEncoding[:0]
+	req.Close = false
+	req.Host = ""
+	resetHTTPMap(req.Form)
+	resetHTTPMap(req.PostForm)
+	req.MultipartForm = nil
+	resetHTTPMap(req.Trailer)
+	req.RemoteAddr = ""
+	req.RequestURI = ""
+	req.TLS = nil
+	// req.Cancel = nil
+	req.Response = nil
+}
+
+func resetHTTPMap(m map[string][]string) {
+	for k := range m {
+		delete(m, k)
+	}
+}
+
+func requestCtxToRequest(ctx *fasthttp.RequestCtx, req *http.Request, httpBody *netHTTPBody) error {
+	body := ctx.Request.Body()
+
+	req.Method = strconv.B2S(ctx.Method())
+	req.Proto = "HTTP/1.1"
+	req.ProtoMajor = 1
+	req.ProtoMinor = 1
+	req.RequestURI = strconv.B2S(ctx.RequestURI())
+	req.ContentLength = int64(len(body))
+	req.Host = strconv.B2S(ctx.Host())
+	req.RemoteAddr = ctx.RemoteAddr().String()
+	req.TLS = ctx.TLSConnectionState()
+
+	httpBody.b = append(httpBody.b[:0], body...)
+	req.Body = httpBody
+
+	req.Header = make(http.Header)
+	ctx.Request.Header.VisitAll(func(k, v []byte) {
+		sk := strconv.B2S(k)
+		sv := strconv.B2S(v)
+
+		switch sk {
+		case "Transfer-Encoding":
+			req.TransferEncoding = append(req.TransferEncoding, sv)
+		default:
+			req.Header.Set(sk, sv)
+		}
+	})
+
+	rURL, err := url.ParseRequestURI(req.RequestURI)
+	if err != nil {
+		return err
+	}
+
+	req.URL = rURL
```

Comment:
They are not the same object type.

Reply:
https://github.com/valyala/fasthttp/blob/master/fasthttpadaptor/adaptor.go#L53

Reply:
> They are not the same object type.

Yeah I know, what I meant was to call `req.URI()` and convert it into a `url.URL`. Something like:

```go
uri := req.URI()
url := &url.URL{
    Host: string(uri.Host()),
    Path: string(uri.Path()),
    ...
}
```

The point being to avoid reparsing the URI, and possibility of an error.